### PR TITLE
2.6 fix intermittent login failure

### PR DESCRIPTION
**Pull request by howbazaar**
This branch fixes two different intermittent failures due to the model not being found in the model cache.
This happens when the apiserver code is creating the facade context.

The fix is to ensure that the model exists in the cache before continuing with the test. A helper method has been added to the JujuConnSuite as this is the one that triggers it most.

Additionally the apiserver is now not created in the engine until the model cache has been initialized. This will help avoid login issues to the apiserver if there are many models.

## QA steps

* bootstrap a controller
* deploy something to ensure the components are hooked up

## Documentation changes

No user facing impact.

### CODE REVIEW

**Comment by howbazaar** [comment]
Address already in use !!build!!

File: worker/modelcache/worker.go [hunk 1]
```diff
@@ -29,8 +29,14 @@ type BackingWatcher interface {
 	Stop() error
 }
 
+// Unlocker is used to indicate that the model cache is ready to be used.
+type Unlocker interface {
+	Unlock()
```
Comment (timClicks) [inline worker/modelcache/worker.go]: Can `Unlock()` fail?

File: worker/modelcache/worker.go [hunk 2]
```diff
@@ -29,8 +29,14 @@ type BackingWatcher interface {
...
+}
+
 // Config describes the necessary fields for NewWorker.
 type Config struct {
+	InitializedGate      Unlocker
```
Comment (timClicks) [inline worker/modelcache/worker.go]: Will tend to stutter as variables are called `gate`.

File: worker/modelcache/worker.go (hunk 1 above)
Comment (howbazaar) [reply worker/modelcache/worker.go]: No.

File: worker/modelcache/worker.go (hunk 2 above)
Comment (howbazaar) [reply worker/modelcache/worker.go]: Actually none of the variables are called `gate`. But `initialized` or `unlocker`. The later of which is a bit odd IMO.

File: juju/testing/conn.go [hunk 3]
```diff
@@ -253,6 +253,28 @@ func (s *JujuConnSuite) WaitForModelWatchersIdle(c *gc.C, modelUUID string) {
 	}
 }
 
+// EnsureCachedModel is used to ensure that the model specified is
+// in the model cache. This is used when tests create models and then
+// want to do things with those models where the actions may touch
+// the model cache.
+func (s *JujuConnSuite) EnsureCachedModel(c *gc.C, uuid string) {
+	start := time.Now()
```
Comment (jameinel) [inline juju/testing/conn.go]: Is this clearer as a

```
stop := time.After(testing.LongWait)
retry := time.After(0)
for {
   select {
   case <-retry:
    _, err := s.Controller.Model(uuid)
    if err == nil {
     return
   }
   timeout = time.After(testing.ShortWait)
   case <-stop:
    c.Errorf("model %v not seen in cache after %v", uuid, testing.LongWait)
    return
 }
}
```

Comment (howbazaar) [reply juju/testing/conn.go]: Yes, I think something with a select is better.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**